### PR TITLE
insightclaw integration in openclaw connector

### PR DESCRIPTION
**Pull request by 91pavan**
Problem
-------

Users installing DefenseClaw on OpenClaw had zero out-of-the-box observability into LLM usage — token counts, tool calls, latency, and cost. InsightClaw (`@outshift-open/insightclaw`) is the plugin that provides this, but it was not part of the automated setup flow. Operators had to install it manually and hand-edit `~/.openclaw/openclaw.json` to wire up the OTel endpoint.

Current Situation
-----------------

`defenseclaw init` (OpenClaw connector path) installed only the DefenseClaw fetch-interceptor plugin. InsightClaw existed as a separate, opt-in step with no documentation in the quickstart and no automated teardown. `VerifyClean` had no awareness of InsightClaw residuals, so a broken partial install could pass clean checks.

The NPM package has since been published as the scoped package `@outshift-open/insightclaw`.

Solution
--------

`installOpenClawExtension` now calls `installInsightClawNPMPlugin` immediately after writing the DefenseClaw plugin files. This runs:

    openclaw plugins install @outshift-open/insightclaw
    

The step is idempotent — if `~/.openclaw/extensions/insightclaw/` already contains a `package.json` or `openclaw.plugin.json`, the install is skipped.

`patchOpenClawConfig` now also manages InsightClaw's config block in one atomic, file-locked write alongside the existing DefenseClaw patching:

*   Appends `insightclaw` to `plugins.allow`
*   Sets `plugins.entries.insightclaw.enabled = true`
*   Seeds `plugins.entries.insightclaw.config` with defaults (endpoint, metrics, traces, spans, etc.) — **only for keys not already present**, so user-set values survive re-runs
*   Adds `~/.openclaw/extensions/insightclaw` to `plugins.load.paths`
*   Records `plugins.installs.insightclaw` (`source: npm`, `sourcePath: @outshift-open/insightclaw`) for OpenClaw's resolver

Teardown (`uninstallOpenClawExtension` and the backup-restore path) now removes the InsightClaw extension directory and purges its entries from `openclaw.json` symmetrically. `VerifyClean` checks for InsightClaw residuals.

A new OTel metrics schema (`schemas/otel/insightclaw-metrics.schema.json`) documents all `openclaw.*` metrics the plugin emits, validated at CI time by `check_schemas.py`.

Architecture Diagram
--------------------

N/A — InsightClaw sits alongside the request path as an OTel emitter, not in it. No change to the proxy or guardrail data flow.

What Changed (Where & How)
--------------------------

| File | Summary |
|------|---------|
| `internal/gateway/connector/openclaw.go` | Added `installInsightClawNPMPlugin`, `openClawPluginDirLooksUsable`, `defaultInsightClawConfig`; added InsightClaw patching in `patchOpenClawConfig`; added InsightClaw cleanup in `uninstallOpenClawExtension`, `Teardown`, and `VerifyClean`; propagated `ctx` into `installOpenClawExtension`. |
| `internal/gateway/connector/connector_test.go` | Added `stubInsightClawInstall` test helper; extended setup, idempotency, HILT, and teardown tests for InsightClaw; added `TestPatchOpenClawConfig_PreservesDefenseClawPluginConfig` to verify user config is not overwritten. |
| `schemas/otel/insightclaw-metrics.schema.json` | New file — OTel metrics catalog documenting all `openclaw.*` instruments emitted by the plugin. |
| `scripts/check_schemas.py` | Added `check_insightclaw_metrics_catalog()` validator; wired it into `main()` so CI fails if the schema is malformed or missing. |
| `docs/GUARDRAIL.md` | Added InsightClaw plugin lifecycle section; updated teardown description and architecture diagram. |
| `docs/GUARDRAIL_QUICKSTART.md` | Updated install description; added InsightClaw to the teardown bullet list; added an InsightClaw troubleshooting section. |
</body></html>

Breaking Changes
----------------

None. `patchOpenClawConfig` only writes InsightClaw config keys that are absent — existing user customisations in `~/.openclaw/openclaw.json` are preserved across reinstalls.

Open Issues / Follow-ups
------------------------

None.

Test Plan
---------

    go test ./internal/gateway/connector/... -v
    

All tests pass, including:

*   `TestOpenClaw_Setup_InstallsExtensionAndPatchesConfig` — verifies InsightClaw is in `plugins.allow`, `plugins.entries`, `plugins.load.paths`, and `plugins.installs`
*   `TestOpenClaw_Setup_IsIdempotent` — verifies running setup twice leaves exactly one InsightClaw entry in each section
*   `TestPatchOpenClawConfig_PreservesDefenseClawPluginConfig` — verifies user-set `endpoint` and `metrics: false` survive a repatch
*   `TestOpenClaw_Teardown_RemovesExtensionAndConfig` — verifies `extensions/insightclaw/` is removed and `plugins.allow` no longer contains `insightclaw`

### CODE REVIEW

File: scripts/check_schemas.py [hunk 1]
```diff
@@ -352,6 +352,128 @@ def check_metrics_catalog(doc: dict) -> bool:
...
+            )
+            ok = False
+
+        if not isinstance(item.get("attributes", {}), dict):
+            print(
+                f"check_schemas: otel/insightclaw-metrics.schema.json: "
+                f"{name}.attributes must be an object",
+                file=sys.stderr,
+            )
+            ok = False
```
[the remark this answers is not in the document]
Comment (91pavan) [reply scripts/check_schemas.py]: fixed

File: internal/gateway/connector/openclaw.go [hunk 2]
```diff
@@ -232,6 +256,40 @@ func installOpenClawExtension(ocHome string, enablePluginApprovals bool) error {
...
+
+func openClawPluginDirLooksUsable(dir string) bool {
+	if _, err := os.Stat(filepath.Join(dir, "openclaw.plugin.json")); err == nil {
+		return true
+	}
+	if _, err := os.Stat(filepath.Join(dir, "package.json")); err == nil {
+		return true
+	}
+	return false
+}
```
[the remark this answers is not in the document]
Comment (91pavan) [reply internal/gateway/connector/openclaw.go]: fixed

**Comment by vineethsai7** [comment]
Security review finding: InsightClaw content capture is enabled by default.

This PR patches the default InsightClaw config with `captureContent: true`. That means a normal DefenseClaw OpenClaw setup can start exporting raw prompt/completion/tool content into telemetry without an explicit operator opt-in. Given DefenseClaw’s redaction-by-default posture for managed telemetry/audit sinks, content-bearing telemetry should be opt-in while metrics/traces remain enabled.

Affected file: `internal/gateway/connector/openclaw.go` (`defaultInsightClawConfig`). The docs also show the same default in `docs/GUARDRAIL.md`.

Recommended fix: set `captureContent` to `false` in the default config, keep preserving explicit user-provided config values, and update the docs to say InsightClaw emits metrics/traces without prompt or completion content by default.

I prepared the narrow patch on branch `codex/fix-pr437-insight-content-default`, but GitHub would not let me open a PR directly against this fork head (`91pavan:feat/insightclaw-integration`). Validation on that branch:

```bash
make sync-openclaw-extension
go test ./internal/gateway/connector -run 'TestOpenClaw_Setup_InstallsExtensionAndPatchesConfig|TestPatchOpenClawConfig_PreservesDefenseClawPluginConfig|TestOpenClaw_Teardown_RemovesExtensionAndConfig'
uv run python scripts/check_schemas.py
```

Observed: focused connector tests pass; schema checks pass.

File: internal/gateway/connector/openclaw.go [hunk 3]
```diff
@@ -232,6 +257,75 @@ func installOpenClawExtension(ocHome string, enablePluginApprovals bool) error {
...
+	if ocHome != filepath.Join(userHomeDir(), ".openclaw") {
+		// Remove any existing OPENCLAW_HOME entry then append the override.
+		filtered := env[:0]
+		for _, e := range env {
+			if !strings.HasPrefix(e, "OPENCLAW_HOME=") {
+				filtered = append(filtered, e)
+			}
+		}
+		env = append(filtered, "OPENCLAW_HOME="+ocHome)
+	}
```
[the remark this answers is not in the document]
Comment (91pavan) [reply internal/gateway/connector/openclaw.go]: fixed

**Comment by vineethsai7** [comment]
Security review finding: `installInsightClawNPMPlugin` installs `@outshift-open/insightclaw` without a version or integrity pin.

That makes normal `defenseclaw setup openclaw` depend on whatever the npm/OpenClaw resolver returns at install time. A registry compromise, maintainer mistake, or breaking latest release can change code that OpenClaw loads without any DefenseClaw release or review boundary. This is especially risky because the install is automatic and runs as part of a guardrail/security setup path.

Affected file: `internal/gateway/connector/openclaw.go` (`insightClawNPMSource` / `openclaw plugins install`).

Recommended fix: pin the package spec to a reviewed version (or vendor the plugin artifact in the DefenseClaw release), record the expected version in `openclaw.json`, and add a setup test that fails if the install source is bare/unversioned.

**Comment by 91pavan** [comment]
> Security review finding: `installInsightClawNPMPlugin` installs `@outshift-open/insightclaw` without a version or integrity pin.
> 
> That makes normal `defenseclaw setup openclaw` depend on whatever the npm/OpenClaw resolver returns at install time. A registry compromise, maintainer mistake, or breaking latest release can change code that OpenClaw loads without any DefenseClaw release or review boundary. This is especially risky because the install is automatic and runs as part of a guardrail/security setup path.
> 
> Affected file: `internal/gateway/connector/openclaw.go` (`insightClawNPMSource` / `openclaw plugins install`).
> 
> Recommended fix: pin the package spec to a reviewed version (or vendor the plugin artifact in the DefenseClaw release), record the expected version in `openclaw.json`, and add a setup test that fails if the install source is bare/unversioned.

Fixed!

**Comment by vineethsai7** [comment]
Security review finding: `policies/default.yaml` now leaves the `defenseclaw` first-party allow-list entry with an empty `source_path_contains` value.

The admission Rego treats a missing/empty provenance list as a match (`not entry.source_path_contains` / `count(...) == 0`), so under the default policy any plugin named `defenseclaw` can bypass scan provenance checks regardless of install path. The intended DefenseClaw path markers were shifted under the new `insightclaw` entry instead.

Affected file: `policies/default.yaml` in `first_party_allow_list`.

Recommended fix: keep the existing `.openclaw/.zeptoclaw/.claude/.codex` DefenseClaw markers under the `target_name: defenseclaw` entry, and give `insightclaw` only its own `.openclaw/extensions/insightclaw` marker. Add a policy parity/admission test that fails when a first-party entry has an empty provenance marker list.

**Comment by vineethsai7** [comment]
Security review finding: a pre-existing local InsightClaw directory is trusted without proving it is the reviewed pinned package.

`installInsightClawNPMPlugin` returns success whenever `~/.openclaw/extensions/insightclaw` contains `openclaw.plugin.json` or `package.json`, before running the pinned `@outshift-open/insightclaw@0.1.3` install. `patchOpenClawConfig` then enables `insightclaw`, adds that directory to `plugins.load.paths`, and records the pinned install metadata even if the files were already attacker-controlled.

Repro: create `~/.openclaw/extensions/insightclaw/package.json` with malicious plugin code before running OpenClaw connector setup. Setup treats the directory as usable and enables it instead of replacing or verifying it.

Recommended fix: verify existing plugin metadata and manifest match the expected package identity/version before accepting the directory, or always reinstall into a freshly controlled directory before enabling it. The test should cover a stale or attacker-precreated `extensions/insightclaw` directory.

**Comment by vineethsai7** [comment]
Daily security review residuals from the July 15 pass:

The high-risk policy provenance and pre-existing InsightClaw directory trust issues are already tracked in earlier comments, so I am not duplicating those here. Two residual items still need cleanup:

1. **Medium: docs still show an unpinned InsightClaw install.** `docs/GUARDRAIL.md` and `docs/GUARDRAIL_QUICKSTART.md` still instruct `openclaw plugins install @outshift-open/insightclaw`, while the code pins `@outshift-open/insightclaw@0.1.3`. Please document the pinned spec or say DefenseClaw setup performs the managed pinned install, and add a doc check so the bare package form does not come back.

2. **Low CI risk: PR head is not gofmt-clean.** `gofmt -l internal/gateway/connector/openclaw.go` reports the file at the current head. Please run gofmt and make sure the relevant formatting check covers this path.

Because this PR is from a fork, I could not open a repair PR directly against the contributor branch from this repository.
